go/plugins/firebase/firebase.go

```diff
@@ -16,31 +16,78 @@ package firebase
 
 import (
 	"context"
+	"fmt"
 	"sync"
 
 	firebase "firebase.google.com/go/v4"
-	"firebase.google.com/go/v4/auth"
 )
 
-type FirebaseApp interface {
-	Auth(ctx context.Context) (*auth.Client, error)
+var state struct {
+	mu      sync.Mutex
+	initted bool
+	app     *firebase.App
 }
 
-var (
-	app   *firebase.App
-	mutex sync.Mutex
-)
+// FirebasePluginConfig is the configuration for the Firebase plugin.
+type FirebasePluginConfig struct {
+	AuthOverride     *map[string]interface{} `json:"databaseAuthVariableOverride"`
+	DatabaseURL      string                  `json:"databaseURL"`
+	ProjectID        string                  `json:"projectId"`
+	ServiceAccountID string                  `json:"serviceAccountId"`
+	StorageBucket    string                  `json:"storageBucket"`
+}
+
+// Init initializes the Firebase app with the provided configuration.
+// If called more than once, it logs a message and returns nil.
+func Init(ctx context.Context, cfg *FirebasePluginConfig) error {
+	state.mu.Lock()
+	defer state.mu.Unlock()
+
+	if state.initted {
+		// throw an error if the app is already initialized
+		return fmt.Errorf("firebase.Init: Firebase app already initialized")
+	}
+
+	// Prepare the Firebase config
+	firebaseConfig := &firebase.Config{
+		AuthOverride:     cfg.AuthOverride,
+		DatabaseURL:      cfg.DatabaseURL,
+		ProjectID:        cfg.ProjectID, // Allow ProjectID to be empty
+		ServiceAccountID: cfg.ServiceAccountID,
+		StorageBucket:    cfg.StorageBucket,
+	}
+
+	// Initialize Firebase app with service account key if provided
+	app, err := firebase.NewApp(ctx, firebaseConfig)
+	if err != nil {
+		return fmt.Errorf("firebase.Init: %w", err)
+	}
 
-// app returns a cached Firebase app.
-func App(ctx context.Context) (FirebaseApp, error) {
-	mutex.Lock()
-	defer mutex.Unlock()
-	if app == nil {
-		newApp, err := firebase.NewApp(ctx, nil)
-		if err != nil {
-			return nil, err
-		}
-		app = newApp
+	state.app = app
+	state.initted = true
+
+	return nil
+}
+
+// UnInit uninitializes the Firebase app.
+
+func UnInit() {
```

Comment:
Test is in the same package, better to drop this and just clean up state directly. We don't want to expose a function like this.